scheduler/scheduler.go

```diff
@@ -69,17 +81,114 @@ type Queue struct {
 	queuedRequestsMetricsKey string
 }
 
+// FifoQueue objects implement a FIFO queue for handling requests,
+// with a maximum allowed concurrency and queue size. Currently, they
+// can be used from the fifo filters in the filters/scheduler package
+// only.
+type FifoQueue struct {
+	queue                    *fifoQueue
+	config                   Config
+	metrics                  metrics.Metrics
+	activeRequestsMetricsKey string
+	errorFullMetricsKey      string
+	errorOtherMetricsKey     string
+	errorTimeoutMetricsKey   string
+	queuedRequestsMetricsKey string
+}
+
+type fifoQueue struct {
+	mu             sync.RWMutex
+	counter        *atomic.Uint64
+	sem            *semaphore.Weighted
+	timeout        time.Duration
+	maxQueueSize   uint64
+	maxConcurrency uint64
+	closed         bool
+}
+
+func (fq *fifoQueue) status() QueueStatus {
+	fq.mu.RLock()
+	maxConcurrency := fq.maxConcurrency
+	fq.mu.RUnlock()
+
+	all := fq.counter.Load()
+
+	var queued, active uint64
+	if all > maxConcurrency {
+		queued = all - maxConcurrency
+		active = maxConcurrency
+	} else {
+		queued = 0
+		active = all
+	}
+
+	return QueueStatus{
+		ActiveRequests: int(active),
+		QueuedRequests: int(queued),
+		Closed:         fq.closed,
+	}
+}
+
+// Reconfigure does nothing, because it creates a data race that would
+// require more mutex than we want to have
+func (fq *fifoQueue) reconfigure(c Config) {
+	fq.mu.Lock()
+	defer fq.mu.Unlock()
+	fq.maxConcurrency = uint64(c.MaxConcurrency)
+	fq.maxQueueSize = uint64(c.MaxQueueSize)
+	fq.timeout = c.Timeout
+}
+
+func (fq *fifoQueue) wait(ctx context.Context) (func(), error) {
+	fq.mu.RLock()
+	maxConcurrency := fq.maxConcurrency
+	maxQueueSize := fq.maxQueueSize
+	timeout := fq.timeout
+	fq.mu.RUnlock()
+
+	// handle queue
+	all := fq.counter.Inc()
+	defer fq.counter.Dec()
+	// queue full?
+	if all > maxConcurrency+maxQueueSize {
+		return func() {}, ErrQueueFull
+	}
+
+	// limit concurrency
+	c, done := context.WithTimeout(ctx, timeout)
+	defer done()
+	if err := fq.sem.Acquire(c, 1); err != nil {
+		switch err {
+		case context.DeadlineExceeded:
+			return func() {}, ErrQueueTimeout
+		case context.Canceled:
+			return func() {}, ErrClientCanceled
+		default:
+			// does not exist yet in Go stdlib as of Go1.18.4
+			return func() {}, err
```

Comment:
I think it should `return nil, ErrQueueTimeout` and caller should not call done() when error is not nil